jgrapht-opt/src/main/java/org/jgrapht/opt/graph/FastutilGraphSpecificsStrategy.java

```diff
@@ -0,0 +1,81 @@
+package org.jgrapht.opt.graph;
+
+import java.io.Serializable;
+import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.function.Supplier;
+
+import org.jgrapht.Graph;
+import org.jgrapht.GraphType;
+import org.jgrapht.graph.GraphSpecificsStrategy;
+import org.jgrapht.graph.specifics.DirectedEdgeContainer;
+import org.jgrapht.graph.specifics.DirectedSpecifics;
+import org.jgrapht.graph.specifics.Specifics;
+import org.jgrapht.graph.specifics.UndirectedEdgeContainer;
+import org.jgrapht.graph.specifics.UndirectedSpecifics;
+
+import it.unimi.dsi.fastutil.objects.Object2ReferenceLinkedOpenHashMap;
+import it.unimi.dsi.fastutil.objects.Object2ReferenceOpenHashMap;
+
+/**
+ * A specifics strategy implementation using fastutil maps for storage.
+ * 
+ * <p>
+ * Graphs constructed using this strategy require the least amount of memory, at the expense of slow
+ * edge retrievals. Methods which depend on edge retrievals, e.g. getEdge(V u, V v), containsEdge(V
+ * u, V v), addEdge(V u, V v), etc may be relatively slow when the average degree of a vertex is
+ * high (dense graphs). For a fast implementation, use
+ * {@link FastutilFastLookupGraphSpecificsStrategy}.
+ * 
+ * @author Dimitrios Michail
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ */
+public class FastutilGraphSpecificsStrategy<V, E>
+    implements
+    GraphSpecificsStrategy<V, E>
+{
+    private static final long serialVersionUID = -4319431062943632549L;
+
+    /**
+     * Get a function which creates the specifics. The factory will accept the graph type as a
+     * parameter.
+     * 
+     * @return a function which creates intrusive edges specifics.
+     */
+    @Override
+    public BiFunction<Graph<V, E>, GraphType, Specifics<V, E>> getSpecificsFactory()
+    {
+        return (BiFunction<Graph<V, E>, GraphType,
+            Specifics<V, E>> & Serializable) (graph, type) -> {
+                if (type.isDirected()) {
+                    return new DirectedSpecifics<>(
+                        graph,
+                        this.<V, DirectedEdgeContainer<V, E>> getPredictableOrderMapFactory().get(),
+                        getEdgeSetFactory());
+                } else {
+                    return new UndirectedSpecifics<>(
+                        graph,
+                        this
+                            .<V, UndirectedEdgeContainer<V, E>> getPredictableOrderMapFactory()
+                            .get(),
+                        getEdgeSetFactory());
+                }
+            };
+    }
+
+    @Override
+    public <K1, V1> Supplier<Map<K1, V1>> getPredictableOrderMapFactory()
+    {
+        return (Supplier<
+            Map<K1, V1>> & Serializable) () -> new Object2ReferenceLinkedOpenHashMap<>();
```

Comment:
Cant we use specific maps based on type of K1? E.g. Int2ReferenceLinkedOpenHashMap if K1 is int? I feel like we are under-utilizing fastutil here.

Reply:
Yes, previously we had talked about having specialized versions for integers etc.  I agree that we don't want to take the generate-all-the-things approach.  Instead, we should start with the most common use cases and make sure we have good coverage for those, then incorporate others over time as it makes sense.

For vertices, integers (or objects which can be mapped to unique integers) make the most sense, probably followed by 64-bit ints.  Then extra specialization for the range of integers [0,n) if we want to support matrix representations.

For edges, I'm not sure what makes sense.  For a matrix representation, when paired with 32-bit integer vertices, a 64-bit int might make sense, with the high 32-bits being one vertex ID and the low 32-bits being the other.

Reply:
Why not just have both this class and the FastUtilMapGraph constructors take an argument of Class<K1> which really is just Class<V> as far as I can tell. Then you know what class V is and can return primitive specific fastutil maps and otherwise default to the Oobject2Reference map.